### PR TITLE
Give preference to routingExternalGWs annotation over hybrid overlay

### LINKED ISSUE #1632: Support external gateway migration from hybrid overlay to OVN
With adding the capability for multiple external gateways (pod or external) using OVN, we need to deprecate the old method using hybrid overlay with vxlan. However for the time being, we need to provide a migration path. In order to limit the amount of downtime for pods in a namespace using external gateways a user needs to:

1. Create external gateways with the new annotations (namespace or pod) that are handled via OVN
2. Respawn pods in any order. This should make the pods get new default routes back to OVN DR (.1) instead of HO br-ext (.3).
3. After all pods are migrated, now remove old HO annotation.

In order to make 2 work, we need to prefer OVN exgws over HO gws. That means if we go to add a pod and any new exgws exist in the namespace, we do not add HO route.

### PR DESCRIPTION
**Pull request by rcarrillocruz**
This is to allow a graceful migration from HO

Fixes #1632

### CODE REVIEW

File: go-controller/pkg/ovn/pods.go
Comment (trozet) [inline go-controller/pkg/ovn/pods.go:183]: you also need to check oc.getRoutingPodGWs(pod.Namespace) right?

Comment (rcarrillocruz) [reply go-controller/pkg/ovn/pods.go:183]: sure

File: test/e2e/e2e_test.go
Comment (aojea) [inline test/e2e/e2e_test.go:203]: if we don't check for an error here we don't know if it failed the check

Comment (aojea) [reply test/e2e/e2e_test.go:203]: this should be 
```suggestion
	err = wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
```

and the chek if err != nil

full disclaimer, the external gateway annotations on pods does not work, it seems the pod never gets the annotation
https://github.com/ovn-org/ovn-kubernetes/pull/1561#discussion_r485782645

**Comment by aojea** [comment]
I'd rather prefer to merge this after the bump and refactor of the control-plane tests
https://github.com/ovn-org/ovn-kubernetes/pull/1561
or we'll keep going with the current technical debt on this area

**Comment by rcarrillocruz** [comment]
> I'd rather prefer to merge this after the bump and refactor of the control-plane tests
> #1561
> or we'll keep going with the current technical debt on this area

yeah, i have no issue waiting. I'll just rebase when that merges.

File: test/e2e/e2e_test.go
Comment (aojea) [reply test/e2e/e2e_test.go:334]: we talked offline, and seems pod does not receive annotations, so we need to check for them on the pod

**Comment by aojea** [comment]
> > I'd rather prefer to merge this after the bump and refactor of the control-plane tests
> > #1561
> > or we'll keep going with the current technical debt on this area
> 
> yeah, i have no issue waiting. I'll just rebase when that merges.

please, disregard my comment and don' t hold on my PR, there are things that has to be fixed in the control-plane e2e tests before I can land it

File: test/scripts/install-kind.sh
Comment (trozet) [inline test/scripts/install-kind.sh:18]: we dont always want to enable hybrid overlay right? Just for e2e control plane tests which should be handled via ENV var?

File: test/e2e/e2e_test.go
Comment (trozet) [inline test/e2e/e2e_test.go:936]: this name is confusing. Just say "routing-external-gw routes to OVN DR should take precedence over hybrid exgw routes to .3"

File: test/e2e/e2e_test.go
Comment (trozet) [inline test/e2e/e2e_test.go:988]: I think you should just change that function to be getPodGWRoute and then compare it here in the test case to see if its .1

File: test/e2e/e2e_test.go
Comment (trozet) [inline test/e2e/e2e_test.go:957]: you dont really need to setup the docker container for this test right, you are just checking annotations.

File: test/e2e/e2e_test.go
Comment (trozet) [inline test/e2e/e2e_test.go:86]: are  all these functions and structs taken from ovn-k8s package? is it possible to import them here or no?

File: test/scripts/install-kind.sh
Comment (rcarrillocruz) [reply test/scripts/install-kind.sh:18]: oops, yeah, that's something I put for testing locally

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:936]: ack

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:957]: possibly, i copied over from previous test so must be it

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:86]: i tried, but ran into all sorts of weird import issues. Showed to @nerdalert and @aojea and we couldn't figure out, so we agreed to just copy it over to the tests.
Maybe land as is and we look at refactoring later on?

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:988]: Oh? I wish I had saw that function earlier...

Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:988]: Ah, you mean changing the semantics of the function, I thought getPodGWRoute was an already existing function that I missed, phew.

File: test/e2e/e2e_test.go
Comment (aojea) [reply test/e2e/e2e_test.go:86]: this folder has its own go modules, it's independent from go-controller.
I don't know if you can play with the redirects in go.mod , but due to the amount of code I think that is easier to copy it directly, but I suggest to create a utils.go file in the same folder with the code, so we don't keep growing this files.

Comment (rcarrillocruz) [reply test/e2e/e2e_test.go:86]: aye, i can decouple on its own file

Comment (trozet) [reply test/e2e/e2e_test.go:86]: sounds good

File: test/e2e/e2e_test.go
Comment (trozet) [reply test/e2e/e2e_test.go]: oh sorry, I meant refactor the checkHybridGWPodRoutes

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go]: done

**Review by trozet — APPROVED** [review]
/lgtm

File: test/e2e/e2e_test.go
Comment (rcarrillocruz) [reply test/e2e/e2e_test.go]: done

**Pull request merged**
